File: Sources/Usage/GeminiCredentials.swift
import Foundation

/// How the Gemini CLI is signed in on this machine. Only the personal-OAuth
/// path (`oauth-personal`) exposes the Code Assist quota API this app reads;
/// api-key / vertex-ai logins have no quota endpoint we can speak to.
enum GeminiAuthDetection: Equatable {
    /// No usable ~/.gemini footprint — zero-intrusion, show nothing.
    case notInstalled
    /// settings.json declares a non-OAuth auth type (api-key, vertex-ai…).
    case unsupportedAuth(String)
    /// oauth_creds.json present on the oauth-personal path.
    case oauthPersonal
}

/// Parsed `~/.gemini/oauth_creds.json`. `expiry_date` is epoch milliseconds
/// (the CLI writes `Date.now() + expires_in * 1000`).
struct GeminiOAuthCreds: Equatable {
    let accessToken: String
    let refreshToken: String?
    let idToken: String?
    let expiryDate: Date?

    var email: String? { idToken.flatMap(GeminiJWT.email(fromIDToken:)) }
}

/// Reader/refresher for the Gemini CLI credential + settings files. Mirrors
/// `GrokAuthFile`'s contract: refresh results MUST be written back (the CLI
/// reads the same file), writes are atomic (tmp file created 0600 in the
/// same directory, then rename(2)), and every field this app doesn't
/// understand is preserved.
enum GeminiCredentials {
    static func homeDirectory() -> URL {
        let home = NSString("~/.gemini").expandingTildeInPath
        return URL(fileURLWithPath: home, isDirectory: true)
    }

    static func settingsURL(home: URL = homeDirectory()) -> URL {
        home.appendingPathComponent("settings.json")
    }

    static func credsURL(home: URL = homeDirectory()) -> URL {
        home.appendingPathComponent("oauth_creds.json")
    }

    /// The configured auth type, or nil when settings.json is missing or
    /// silent about it (the CLI defaults to oauth-personal in that case).
    /// Accepts the three spellings the CLI has shipped: top-level
    /// `selectedAuthType` (classic), top-level `authType`, and the nested
    /// `security.auth.selectedType` (current).
    static func authType(fromSettings data: Data) -> String? {
        guard let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        if let direct = nonEmpty(root["selectedAuthType"]) { return direct }
        if let direct = nonEmpty(root["authType"]) { return direct }
        guard let security = root["security"] as? [String: Any],
              let auth = security["auth"] as? [String: Any] else { return nil }
        return nonEmpty(auth["selectedType"])
    }

    static func detect(home: URL = homeDirectory()) -> GeminiAuthDetection {
        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(atPath: home.path, isDirectory: &isDir),
              isDir.boolValue else { return .notInstalled }
        if let data = try? Data(contentsOf: settingsURL(home: home)),
           let type = authType(fromSettings: data),
           type != "oauth-personal" {
            return .unsupportedAuth(type)
        }
        // A bare ~/.gemini from an aborted install has nothing to show.
        guard FileManager.default.fileExists(atPath: credsURL(home: home).path) else {
            return .notInstalled
        }
        return .oauthPersonal
    }

    static func loadCreds(from url: URL) -> GeminiOAuthCreds? {
        guard let data = try? Data(contentsOf: url),
              let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let accessToken = nonEmpty(root["access_token"]) else { return nil }
        return GeminiOAuthCreds(
            accessToken: accessToken,
            refreshToken: nonEmpty(root["refresh_token"]),
            idToken: nonEmpty(root["id_token"]),
            expiryDate: parseExpiry(root["expiry_date"])
        )
    }

    /// Refresh a minute early so an in-flight request never races the clock.
    static func needsRefresh(_ creds: GeminiOAuthCreds,
                             now: Date = Date(),
                             skew: TimeInterval = 60) -> Bool {
        guard let expiryDate = creds.expiryDate else { return false }
        return now.addingTimeInterval(skew) >= expiryDate
    }

    /// Apply a Google token response (`access_token` / `expires_in` seconds,
    /// occasionally a fresh `id_token`) to oauth_creds.json and persist.
    /// Returns the updated creds, or nil when nothing was written — the
    /// caller keeps using its in-memory token.
    @discardableResult
    static func applyRefreshResponse(_ data: Data,
                                     to url: URL,
                                     now: Date = Date()) -> GeminiOAuthCreds? {
        guard let response = try? JSONSerialization.jsonObject(with: data) as? [String: Any],
              let accessToken = nonEmpty(response["access_token"]) else { return nil }

        guard let fileData = try? Data(contentsOf: url),
              var root = try? JSONSerialization.jsonObject(with: fileData) as? [String: Any] else {
            return nil
        }

        root["access_token"] = accessToken
        if let expiresIn = seconds(response["expires_in"]), expiresIn > 0 {
            root["expiry_date"] = Int((now.timeIntervalSince1970 + expiresIn) * 1000)
        }
        if let rotatedRefresh = nonEmpty(response["refresh_token"]) {
            root["refresh_token"] = rotatedRefresh
        }
        if let freshIDToken = nonEmpty(response["id_token"]) {
            root["id_token"] = freshIDToken
        }

        guard let output = try? JSONSerialization.data(
            withJSONObject: root, options: [.prettyPrinted, .sortedKeys]
        ) else { return nil }

        let tmp = url.deletingLastPathComponent()
            .appendingPathComponent(".oauth_creds.json.tmp-\(UUID().uuidString)")
        guard FileManager.default.createFile(
            atPath: tmp.path, contents: output,
            attributes: [.posixPermissions: 0o600]
        ) else { return nil }
        guard rename(tmp.path, url.path) == 0 else {
            try? FileManager.default.removeItem(at: tmp)
            return nil
        }
        return GeminiOAuthCreds(
            accessToken: accessToken,
            refreshToken: nonEmpty(root["refresh_token"]),
            idToken: nonEmpty(root["id_token"]),
            expiryDate: parseExpiry(root["expiry_date"])
        )
    }

    /// `expiry_date` is epoch milliseconds on current CLI logins; tolerate
    /// epoch seconds from other writer versions.
    static func parseExpiry(_ value: Any?) -> Date? {
        if let raw = value as? Double {
            return Date(timeIntervalSince1970: raw > 1e11 ? raw / 1000 : raw)
        }
        if let raw = value as? Int {
            return parseExpiry(Double(raw))
        }
        return nil
    }

    private static func seconds(_ value: Any?) -> TimeInterval? {
        if let raw = value as? Double { return raw }
        if let raw = value as? Int { return TimeInterval(raw) }
        if let raw = value as? String { return TimeInterval(raw) }
        return nil
    }

    private static func nonEmpty(_ value: Any?) -> String? {
        guard let raw = value as? String, !raw.isEmpty else { return nil }
        return raw
    }
}

/// Minimal JWT payload reader — enough to pull the account email out of the
/// Google `id_token` without any signature verification (we only display it).
enum GeminiJWT {
    static func payload(fromJWT jwt: String) -> [String: Any]? {
        let segments = jwt.split(separator: ".")
        guard segments.count >= 2 else { return nil }
        var base64 = String(segments[1])
            .replacingOccurrences(of: "-", with: "+")
            .replacingOccurrences(of: "_", with: "/")
        while base64.count % 4 != 0 { base64.append("=") }
        guard let data = Data(base64Encoded: base64),
              let object = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
            return nil
        }
        return object
    }

    static func email(fromIDToken idToken: String) -> String? {
        guard let email = payload(fromJWT: idToken)?["email"] as? String,
              !email.isEmpty else { return nil }
        return email
    }
}

/// Google OAuth client id/secret for the token-refresh call. Google ships
/// them inside the Gemini CLI itself (`oauth2.js`), so we extract them from
/// the local install at runtime instead of hardcoding another product's
/// credentials into this binary.
struct GeminiClientCredentials: Equatable {
    let clientID: String
    let clientSecret: String
}

enum GeminiClientExtractor {
    static let clientIDEnvKey = "GEMINI_OAUTH_CLIENT_ID"
    static let clientSecretEnvKey = "GEMINI_OAUTH_CLIENT_SECRET"
    private static let binaryEnvKeys = ["GEMINI_PATH", "GEMINI_CLI_PATH"]
    private static let coreRelativePath =
        "node_modules/@google/gemini-cli-core/dist/src/code_assist/oauth2.js"

    /// Regex pull of `OAUTH_CLIENT_ID` / `OAUTH_CLIENT_SECRET` from the CLI's
    /// oauth2.js (or any bundle chunk that inlines the same constants).
    static func extract(fromOAuth2JS content: String) -> GeminiClientCredentials? {
        guard let id = firstMatch(#"OAUTH_CLIENT_ID\s*=\s*["']([^"']+)["']"#, in: content),
              let secret = firstMatch(#"OAUTH_CLIENT_SECRET\s*=\s*["']([^"']+)["']"#, in: content)
        else { return nil }
        return GeminiClientCredentials(clientID: id, clientSecret: secret)
    }

    static func fromEnvironment(
        _ env: [String: String] = ProcessInfo.processInfo.environment
    ) -> GeminiClientCredentials? {
        guard let id = env[clientIDEnvKey]?.trimmingCharacters(in: .whitespacesAndNewlines),
              let secret = env[clientSecretEnvKey]?.trimmingCharacters(in: .whitespacesAndNewlines),
              !id.isEmpty, !secret.isEmpty else { return nil }
        return GeminiClientCredentials(clientID: id, clientSecret: secret)
    }

    /// Env override first, then the local gemini-cli install.
    static func resolve() -> GeminiClientCredentials? {
        if let fromEnv = fromEnvironment() { return fromEnv }
        guard let jsURL = locateOAuth2JS(),
              let content = try? String(contentsOf: jsURL, encoding: .utf8) else { return nil }
        return extract(fromOAuth2JS: content)
    }

    /// Walk up from the resolved `gemini` binary looking for the CLI package
    /// in its three shipped layouts (direct core install, npm-style nested
    /// package, Homebrew libexec), then fall back to the fixed Homebrew
    /// package roots for setups where the binary itself isn't findable.
    static func locateOAuth2JS(
        env: [String: String] = ProcessInfo.processInfo.environment
    ) -> URL? {
        if let binary = locateGeminiBinary(env: env),
           let near = oauth2JS(nearBinary: binary) {
            return near
        }
        let home = NSString("~").expandingTildeInPath
        let packageRoots = [
            "/opt/homebrew/opt/gemini-cli/libexec/lib/node_modules/@google/gemini-cli",
            "/usr/local/opt/gemini-cli/libexec/lib/node_modules/@google/gemini-cli",
            "\(home)/.local/lib/node_modules/@google/gemini-cli",
        ]
        for root in packageRoots {
            let candidate = URL(fileURLWithPath: root).appendingPathComponent(coreRelativePath)
            if FileManager.default.fileExists(atPath: candidate.path) { return candidate }
        }
        return nil
    }

    static func locateGeminiBinary(
        env: [String: String] = ProcessInfo.processInfo.environment
    ) -> URL? {
        for key in binaryEnvKeys {
            if let raw = env[key], !raw.isEmpty,
               isExecutableFile(raw) {
                return URL(fileURLWithPath: raw)
            }
        }
        if let pathEnv = env["PATH"] {
            for dir in pathEnv.split(separator: ":") where !dir.isEmpty {
                let candidate = "\(dir)/gemini"
                if isExecutableFile(candidate) { return URL(fileURLWithPath: candidate) }
            }
        }
        let home = NSString("~").expandingTildeInPath
        let candidates = [
            "\(home)/.local/bin/gemini",
            "/opt/homebrew/bin/gemini",
            "/usr/local/bin/gemini",
            "/usr/bin/gemini",
        ]
        for candidate in candidates where isExecutableFile(candidate) {
            return URL(fileURLWithPath: candidate)
        }
        return nil
    }

    private static func oauth2JS(nearBinary binary: URL) -> URL? {
        let resolved = binary.resolvingSymlinksInPath()
        var dir = resolved.deletingLastPathComponent()
        for _ in 0..<8 {
            let relatives = [
                coreRelativePath,
                "@google/gemini-cli/\(coreRelativePath)",
                "lib/node_modules/@google/gemini-cli/\(coreRelativePath)",
                "libexec/lib/node_modules/@google/gemini-cli/\(coreRelativePath)",
            ]
            for relative in relatives {
                let candidate = dir.appendingPathComponent(relative)
                if FileManager.default.fileExists(atPath: candidate.path) { return candidate }
            }
            let parent = dir.deletingLastPathComponent()
            if parent.path == dir.path { break }
            dir = parent
        }
        return nil
    }

    private static func isExecutableFile(_ path: String) -> Bool {
        var isDir: ObjCBool = false
        guard FileManager.default.fileExists(atPath: path, isDirectory: &isDir),
              !isDir.boolValue else { return false }
        return FileManager.default.isExecutableFile(atPath: path)
    }

    private static func firstMatch(_ pattern: String, in content: String) -> String? {
        guard let regex = try? NSRegularExpression(pattern: pattern),
              let match = regex.firstMatch(
                in: content,
                range: NSRange(content.startIndex..., in: content)
              ),
              match.numberOfRanges >= 2,
              let range = Range(match.range(at: 1), in: content) else { return nil }
        let value = String(content[range])
        return value.isEmpty ? nil : value
    }
}

File: windows/src/AgentIsland/Usage/GeminiCredentials.cs
using System.Globalization;
using System.IO;
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;
using AgentIsland.Core;

namespace AgentIsland.Usage;

/// How the Gemini CLI is signed in on this machine. Only the personal-OAuth
/// path (`oauth-personal`) exposes the Code Assist quota API this app reads;
/// api-key / vertex-ai logins have no quota endpoint we can speak to, and the
/// UI must say so rather than spin as if data were coming.
public abstract record GeminiAuthDetection
{
    /// No usable ~\.gemini footprint — zero-intrusion, show nothing.
    public sealed record NotInstalled : GeminiAuthDetection;

    /// settings.json declares a non-OAuth auth type (api-key, vertex-ai…).
    public sealed record UnsupportedAuth(string AuthType) : GeminiAuthDetection;

    /// oauth_creds.json present on the oauth-personal path.
    public sealed record OauthPersonal : GeminiAuthDetection;
}

/// Parsed `%USERPROFILE%\.gemini\oauth_creds.json`. `expiry_date` is epoch
/// milliseconds (the CLI writes `Date.now() + expires_in * 1000`).
public sealed record GeminiOAuthCreds(
    string AccessToken,
    string? RefreshToken,
    string? IdToken,
    DateTimeOffset? ExpiryDate)
{
    public string? Email => IdToken is { } token ? GeminiJwt.Email(token) : null;
}

/// Reader/refresher for the Gemini CLI credential + settings files. Refresh
/// results MUST be written back (the CLI reads the same file), writes are
/// atomic (temp file in the same directory, then a replacing move), and every
/// field this app doesn't understand is preserved.
public static class GeminiCredentials
{
    public static string HomeDirectory => Path.Combine(IslandPaths.Home, ".gemini");

    public static string SettingsFile => Path.Combine(HomeDirectory, "settings.json");

    public static string CredsFile => Path.Combine(HomeDirectory, "oauth_creds.json");

    /// The gemini-cli tolerates comments and trailing commas in its own
    /// settings file, so a hand-edited settings.json must not read as "no
    /// auth type declared" — that would silently downgrade an api-key login
    /// to "oauth-personal" and imply live quota we cannot fetch.
    private static readonly JsonDocumentOptions SettingsJsonOptions = new()
    {
        AllowTrailingCommas = true,
        CommentHandling = JsonCommentHandling.Skip,
    };

    private static readonly JsonSerializerOptions WriteOptions = new() { WriteIndented = true };

    /// The configured auth type, or null when settings.json is missing or
    /// silent about it (the CLI defaults to oauth-personal in that case).
    /// Accepts the three spellings the CLI has shipped: top-level
    /// `selectedAuthType` (classic), top-level `authType`, and the nested
    /// `security.auth.selectedType` (current).
    public static string? AuthTypeFromSettings(byte[] data)
    {
        try
        {
            using var doc = JsonDocument.Parse(data, SettingsJsonOptions);
            var root = doc.RootElement;
            if (NonEmpty(Jsonl.GetString(root, "selectedAuthType")) is { } classic) return classic;
            if (NonEmpty(Jsonl.GetString(root, "authType")) is { } direct) return direct;
            if (Jsonl.GetObject(root, "security") is not { } security) return null;
            if (Jsonl.GetObject(security, "auth") is not { } auth) return null;
            return NonEmpty(Jsonl.GetString(auth, "selectedType"));
        }
        catch
        {
            return null;
        }
    }

    public static GeminiAuthDetection Detect()
    {
        try
        {
            if (!Directory.Exists(HomeDirectory)) return new GeminiAuthDetection.NotInstalled();
            var settings = SettingsFile;
            if (File.Exists(settings)
                && AuthTypeFromSettings(ReadShared(settings)) is { } type
                && type != "oauth-personal")
            {
                return new GeminiAuthDetection.UnsupportedAuth(type);
            }
            // A bare ~\.gemini from an aborted install has nothing to show.
            if (!File.Exists(CredsFile)) return new GeminiAuthDetection.NotInstalled();
            return new GeminiAuthDetection.OauthPersonal();
        }
        catch
        {
            return new GeminiAuthDetection.NotInstalled();
        }
    }

    public static GeminiOAuthCreds? LoadCreds(string path)
    {
        try
        {
            if (!File.Exists(path)) return null;
            using var stream = OpenShared(path);
            using var doc = JsonDocument.Parse(stream);
            var root = doc.RootElement;
            if (NonEmpty(Jsonl.GetString(root, "access_token")) is not { } accessToken) return null;
            return new GeminiOAuthCreds(
                accessToken,
                NonEmpty(Jsonl.GetString(root, "refresh_token")),
                NonEmpty(Jsonl.GetString(root, "id_token")),
                ParseExpiry(root, "expiry_date"));
        }
        catch
        {
            return null;
        }
    }

    /// Refresh a minute early so an in-flight request never races the clock.
    public static bool NeedsRefresh(GeminiOAuthCreds creds, DateTimeOffset? now = null, double skewSeconds = 60)
    {
        if (creds.ExpiryDate is not { } expiry) return false;
        return (now ?? DateTimeOffset.Now).AddSeconds(skewSeconds) >= expiry;
    }

    /// Apply a Google token response (`access_token` / `expires_in` seconds,
    /// occasionally a fresh `id_token`) to oauth_creds.json and persist.
    /// Returns the updated creds, or null when nothing was written — the
    /// caller keeps using its in-memory token.
    public static GeminiOAuthCreds? ApplyRefreshResponse(byte[] data, string path, DateTimeOffset? now = null)
    {
        try
        {
            using var doc = JsonDocument.Parse(data);
            var response = doc.RootElement;
            if (NonEmpty(Jsonl.GetString(response, "access_token")) is not { } accessToken) return null;

            if (!File.Exists(path)) return null;
            JsonObject root;
            using (var existing = OpenShared(path))
            {
                if (JsonNode.Parse(existing) is not JsonObject parsed) return null;
                root = parsed;
            }

            // Track the surviving values in locals: reading them back out of
            // the nodes we just assigned would depend on JsonValue's CLR-type
            // round-tripping rules, which differ from the parsed-file case.
            var refreshToken = NodeString(root["refresh_token"]);
            var idToken = NodeString(root["id_token"]);
            var expiry = ParseExpiry(root["expiry_date"]);

            root["access_token"] = accessToken;
            if (Seconds(response, "expires_in") is { } expiresIn && expiresIn > 0)
            {
                expiry = (now ?? DateTimeOffset.Now).AddSeconds(expiresIn);
                root["expiry_date"] = expiry.Value.ToUnixTimeMilliseconds();
            }
            if (NonEmpty(Jsonl.GetString(response, "refresh_token")) is { } rotated)
            {
                refreshToken = rotated;
                root["refresh_token"] = rotated;
            }
            if (NonEmpty(Jsonl.GetString(response, "id_token")) is { } freshIdToken)
            {
                idToken = freshIdToken;
                root["id_token"] = freshIdToken;
            }

            if (!WriteAtomically(path, root.ToJsonString(WriteOptions))) return null;
            return new GeminiOAuthCreds(accessToken, refreshToken, idToken, expiry);
        }
        catch
        {
            return null;
        }
    }

    /// The gemini CLI keeps oauth_creds.json open while it runs; the default
    /// File.ReadAll* share mode throws against that, and a failed read here
    /// reads to the caller as "never signed in" — or, in the writeback path,
    /// silently drops the rotated refresh token Google already invalidated.
    private static FileStream OpenShared(string path) => new(
        path, FileMode.Open, FileAccess.Read, FileShare.ReadWrite | FileShare.Delete);

    private static byte[] ReadShared(string path)
    {
        using var stream = OpenShared(path);
        using var memory = new MemoryStream();
        stream.CopyTo(memory);
        return memory.ToArray();
    }

    /// Temp file beside the target, then a replacing move. Windows has no
    /// posix mode bits to set — the file inherits the ACL of %USERPROFILE%,
    /// which is already per-user. The move is retried because the CLI may
    /// hold a brief read lock, and a dropped writeback strands a rotated
    /// refresh token that Google has already invalidated server-side.
    private static bool WriteAtomically(string path, string contents)
    {
        var tmp = path + ".tmp-" + Guid.NewGuid().ToString("N");
        try
        {
            File.WriteAllText(tmp, contents);
            for (var attempt = 0; ; attempt++)
            {
                try
                {
                    File.Move(tmp, path, overwrite: true);
                    return true;
                }
                catch (IOException) when (attempt < 4)
                {
                    System.Threading.Thread.Sleep(40 * (attempt + 1));
                }
            }
        }
        catch (Exception error)
        {
            System.Diagnostics.Debug.WriteLine($"AgentIsland: failed to write Gemini creds: {error.Message}");
            try { File.Delete(tmp); } catch { }
            return false;
        }
    }

    /// `expiry_date` is epoch milliseconds on current CLI logins; tolerate
    /// epoch seconds from other writer versions.
    public static DateTimeOffset? ParseExpiry(double raw)
    {
        try
        {
            return raw > 1e11
                ? DateTimeOffset.FromUnixTimeMilliseconds((long)raw)
                : DateTimeOffset.FromUnixTimeSeconds((long)raw);
        }
        catch
        {
            return null;
        }
    }

    private static DateTimeOffset? ParseExpiry(JsonElement root, string property)
    {
        if (root.ValueKind != JsonValueKind.Object) return null;
        if (!root.TryGetProperty(property, out var value)) return null;
        if (value.ValueKind != JsonValueKind.Number) return null;
        return value.TryGetDouble(out var raw) ? ParseExpiry(raw) : null;
    }

    private static DateTimeOffset? ParseExpiry(JsonNode? node)
    {
        if (node is not JsonValue value) return null;
        return value.TryGetValue<double>(out var raw) ? ParseExpiry(raw) : null;
    }

    private static string? NodeString(JsonNode? node)
    {
        if (node is not JsonValue value) return null;
        return value.TryGetValue<string>(out var text) ? NonEmpty(text) : null;
    }

    private static double? Seconds(JsonElement root, string property)
    {
        if (root.ValueKind != JsonValueKind.Object) return null;
        if (!root.TryGetProperty(property, out var value)) return null;
        return value.ValueKind switch
        {
            JsonValueKind.Number => value.TryGetDouble(out var number) ? (double?)number : null,
            JsonValueKind.String => double.TryParse(
                value.GetString(), NumberStyles.Float, CultureInfo.InvariantCulture, out var parsed)
                ? (double?)parsed
                : null,
            _ => null,
        };
    }

    private static string? NonEmpty(string? raw) => string.IsNullOrEmpty(raw) ? null : raw;
}

/// Minimal JWT payload reader — enough to pull the account email out of the
/// Google `id_token` without any signature verification (we only display it).
public static class GeminiJwt
{
    private static JsonDocument? Payload(string jwt)
    {
        var segments = jwt.Split('.');
        if (segments.Length < 2) return null;
        var base64 = segments[1].Replace('-', '+').Replace('_', '/');
        switch (base64.Length % 4)
        {
            case 2: base64 += "=="; break;
            case 3: base64 += "="; break;
            // A one-over remainder is not a valid base64url payload.
            case 1: return null;
        }
        try
        {
            return JsonDocument.Parse(Convert.FromBase64String(base64));
        }
        catch
        {
            return null;
        }
    }

    public static string? Email(string idToken)
    {
        using var payload = Payload(idToken);
        if (payload is null) return null;
        var email = Jsonl.GetString(payload.RootElement, "email");
        return string.IsNullOrEmpty(email) ? null : email;
    }
}

/// Google OAuth client id/secret for the token-refresh call. Google ships
/// them inside the Gemini CLI itself (`oauth2.js`), so we extract them from
/// the local install at runtime instead of hardcoding another product's
/// credentials into this binary.
public sealed record GeminiClientCredentials(string ClientId, string ClientSecret);

public static class GeminiClientExtractor
{
    public const string ClientIdEnvKey = "GEMINI_OAUTH_CLIENT_ID";
    public const string ClientSecretEnvKey = "GEMINI_OAUTH_CLIENT_SECRET";

    private static readonly string[] BinaryEnvKeys = { "GEMINI_PATH", "GEMINI_CLI_PATH" };

    private const string PackageScope = "@google";
    private const string PackageName = "gemini-cli";

    /// Declared before RelativeLayouts: that initializer reads this one, and
    /// static field initializers run in declaration order.
    private static readonly string CoreRelativePath = Path.Combine(
        "node_modules", PackageScope, "gemini-cli-core", "dist", "src", "code_assist", "oauth2.js");

    /// The layouts npm has shipped, walking up from wherever the `gemini`
    /// shim landed: hoisted core, the nested package under a node_modules
    /// root, the scope directory itself, and the unix-style prefix that
    /// MSYS/Git-Bash npm installs still use.
    private static readonly string[] RelativeLayouts =
    {
        CoreRelativePath,
        Path.Combine("node_modules", PackageScope, PackageName, CoreRelativePath),
        Path.Combine(PackageScope, PackageName, CoreRelativePath),
        Path.Combine("lib", "node_modules", PackageScope, PackageName, CoreRelativePath),
    };

    private static readonly Regex ClientIdPattern = new(
        @"OAUTH_CLIENT_ID\s*=\s*[""']([^""']+)[""']", RegexOptions.None, TimeSpan.FromSeconds(2));

    private static readonly Regex ClientSecretPattern = new(
        @"OAUTH_CLIENT_SECRET\s*=\s*[""']([^""']+)[""']", RegexOptions.None, TimeSpan.FromSeconds(2));

    /// Regex pull of `OAUTH_CLIENT_ID` / `OAUTH_CLIENT_SECRET` from the CLI's
    /// oauth2.js (or any bundle chunk that inlines the same constants).
    public static GeminiClientCredentials? Extract(string content)
    {
        if (FirstMatch(ClientIdPattern, content) is not { } id) return null;
        if (FirstMatch(ClientSecretPattern, content) is not { } secret) return null;
        return new GeminiClientCredentials(id, secret);
    }

    public static GeminiClientCredentials? FromEnvironment()
    {
        var id = Environment.GetEnvironmentVariable(ClientIdEnvKey)?.Trim();
        var secret = Environment.GetEnvironmentVariable(ClientSecretEnvKey)?.Trim();
        if (string.IsNullOrEmpty(id) || string.IsNullOrEmpty(secret)) return null;
        return new GeminiClientCredentials(id, secret);
    }

    /// Env override first, then the local gemini-cli install.
    public static GeminiClientCredentials? Resolve()
    {
        if (FromEnvironment() is { } fromEnv) return fromEnv;
        if (LocateOAuth2Js() is not { } path) return null;
        try
        {
            return Extract(File.ReadAllText(path));
        }
        catch
        {
            return null;
        }
    }

    /// Walk up from the resolved `gemini` shim looking for the CLI package,
    /// then fall back to the fixed package roots for setups where the shim
    /// itself isn't findable.
    public static string? LocateOAuth2Js()
    {
        try
        {
            if (LocateGeminiBinary() is { } binary && OAuth2JsNearBinary(binary) is { } near) return near;
            foreach (var root in PackageRoots())
            {
                var candidate = Path.Combine(root, CoreRelativePath);
                if (File.Exists(candidate)) return candidate;
            }
        }
        catch
        {
        }
        return null;
    }

    /// The env overrides, then the shared CLI locator — it already probes
    /// every npm/pnpm/Volta/nvm shim directory plus PATH.
    public static string? LocateGeminiBinary()
    {
        foreach (var key in BinaryEnvKeys)
        {
            var raw = Environment.GetEnvironmentVariable(key);
            if (string.IsNullOrWhiteSpace(raw)) continue;
            if (File.Exists(raw)) return raw;
        }
        return Trigger.CLILocator.Locate("gemini");
    }

    private static string? OAuth2JsNearBinary(string binary)
    {
        string? dir;
        try
        {
            dir = Path.GetDirectoryName(Path.GetFullPath(binary));
        }
        catch
        {
            return null;
        }
        for (var depth = 0; depth < 8 && !string.IsNullOrEmpty(dir); depth++)
        {
            foreach (var relative in RelativeLayouts)
            {
                var candidate = Path.Combine(dir, relative);
                if (File.Exists(candidate)) return candidate;
            }
            var parent = Path.GetDirectoryName(dir);
            if (parent is null || parent == dir) break;
            dir = parent;
        }
        return null;
    }

    private static IEnumerable<string> PackageRoots()
    {
        var roots = new List<string>
        {
            // npm's default global prefix on Windows.
            Path.Combine(IslandPaths.RoamingAppData, "npm", "node_modules", PackageScope, PackageName),
        };

        foreach (var variable in new[] { "npm_config_prefix", "NPM_CONFIG_PREFIX" })
        {
            var prefix = Environment.GetEnvironmentVariable(variable);
            if (string.IsNullOrWhiteSpace(prefix)) continue;
            roots.Add(Path.Combine(prefix, "node_modules", PackageScope, PackageName));
        }

        var programFiles = Environment.GetFolderPath(Environment.SpecialFolder.ProgramFiles);
        if (!string.IsNullOrEmpty(programFiles))
        {
            roots.Add(Path.Combine(programFiles, "nodejs", "node_modules", PackageScope, PackageName));
        }

        // pnpm's global installs sit under a store-version directory that the
        // shim directory does not contain, so walking up from the shim can
        // never reach them.
        var pnpmGlobal = Path.Combine(IslandPaths.LocalAppData, "pnpm", "global");
        try
        {
            if (Directory.Exists(pnpmGlobal))
            {
                foreach (var version in Directory.EnumerateDirectories(pnpmGlobal))
                {
                    roots.Add(Path.Combine(version, "node_modules", PackageScope, PackageName));
                }
            }
        }
        catch
        {
        }

        roots.Add(Path.Combine(IslandPaths.Home, ".local", "lib", "node_modules", PackageScope, PackageName));
        return roots;
    }

    private static string? FirstMatch(Regex pattern, string content)
    {
        try
        {
            var match = pattern.Match(content);
            if (!match.Success || match.Groups.Count < 2) return null;
            var value = match.Groups[1].Value;
            return value.Length == 0 ? null : value;
        }
        catch (RegexMatchTimeoutException)
        {
            return null;
        }
    }
}
